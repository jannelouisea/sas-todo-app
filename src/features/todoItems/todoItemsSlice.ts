import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import {
    IToDoItem
} from 'src/interfaces'
import moment from 'moment';
import {
    Sort
} from 'src/static/enums'

export type ITodoItemAddedPayload = {
    id: string,
    text: string,
    createdAt: number,
};

const initialState: IToDoItem[] = [
    {
        id: 'init-item-001',
        text: 'Clean the house',
        createdAt: moment().unix(),
        completed: false
    },
    {
        id: 'init-item-002',
        text: 'Pack things for upcoming trip',
        createdAt: 1689652800, // Tue Jul 18 2023 00:00:00 GMT-0400 (EST)
        completed: false
    },
    {
        id: 'init-item-003',
        text: 'Make dinner',
        createdAt: 1690257600, // Tue Jul 25 2023 00:00:00 GMT-0400 (EST)
        completed: true
    },
    {
        id: 'init-item-004',
        text: 'Prepare for SAS Interview',
        createdAt: 1689393600, // Sat Jul 15 2023 00:00:00 GMT-0400 (EST)

        completed: true
    },
];

const todoItemsSlice = createSlice({
    name: 'todoItems',
    initialState,
    reducers: {
        todoItemAdded(state, action: PayloadAction<ITodoItemAddedPayload>) {
            const { id, text, createdAt } = action.payload;
            const newItem = {
                id,
                text,
                createdAt,
                completed: false
            };

            return [newItem, ...state];
        },
        todoItemToggled(state, action: PayloadAction<string>) {
            return state.map(item => {
                if (item.id !== action.payload) {
                    return item
                }

                return {
                    ...item,
                    completed: !item.completed
                }
            });
        },
        todoItemRemoved(state, action: PayloadAction<string>) {
            return state.filter(item => item.id !== action.payload);
        },
        todoItemsSorted(state, action: PayloadAction<Sort>) {
            let newState = [...state];

            switch (action.payload) {
                case Sort.DateDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.createdAt - itemA.createdAt
                        //moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                    );
                case Sort.DateAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.createdAt - itemB.createdAt
                        //moment(itemA.createdAt).format('YYYYMMDD').localeCompare(moment(itemB.createdAt).format('YYYYMMDD'))
                    );
                case Sort.TextDesc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemB.text.localeCompare(itemA.text)
                    );
                case Sort.TextAsc:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) =>
                        itemA.text.localeCompare(itemB.text)
                    );
                case Sort.Cmpl1st:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return -1;
                        } else {
                            return 1;
                        }
                    });
                case Sort.Incmpl1st:
                default:
                    return newState.sort((itemA: IToDoItem, itemB: IToDoItem) => {
                        if (itemA.completed === itemB.completed) {
                            //return moment(itemB.createdAt).format('YYYYMMDD').localeCompare(moment(itemA.createdAt).format('YYYYMMDD'))
                            return itemB.createdAt - itemA.createdAt;
                        } else if (itemA.completed) {
                            return 1;
                        } else {
                            return -1;
                        }
                    });
            }
        }
    }
});

export const { todoItemAdded, todoItemToggled, todoItemRemoved, todoItemsSorted } = todoItemsSlice.actions;

export default todoItemsSlice.reducer;