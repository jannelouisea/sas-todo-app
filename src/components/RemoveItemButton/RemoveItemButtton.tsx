import IconButton from '@mui/joy/IconButton';
import Delete from '@mui/icons-material/Delete';
import Tooltip from '@mui/joy/Tooltip';

import { useAppDispatch } from 'src/app/hooks';
import { todoItemRemoved } from 'src/features/todoItems/todoItemsSlice';

import {
    MUIColor,
    MUISize,
    MUIVariant,
    MUIPos,
    MUIIconColor
} from 'src/static/enums'

type Props = {
    itemId: string,
    color: MUIColor
}

function RemoveItemButtton({ itemId, color }: Props) {
    const dispatch = useAppDispatch();

    const onRemove = () => {
        dispatch(todoItemRemoved(itemId));
    }

    return (
        <Tooltip title='Remove item' size={MUISize.Medium} variant={MUIVariant.Outlined} placement={MUIPos.Right}>
            <IconButton
                id='remove-item-button'
                aria-label='Remove item button'
                onClick={onRemove}
                size={MUISize.Small}
                variant={MUIVariant.Plain}
                color={color}
            >
                <Delete sx={{ color: MUIIconColor.Inherit }} />
            </IconButton>
        </Tooltip>
    );
}

export default RemoveItemButtton;