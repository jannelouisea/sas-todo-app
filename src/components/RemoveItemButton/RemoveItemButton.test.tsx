import { Provider } from 'react-redux';
import {
    IStoreType,
    createStore
} from 'src/app/store';
import RemoveItemButton from './RemoveItemButton';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import {
    MUIColor,
} from 'src/static/enums';
import * as todoItemsActions from 'src/features/todoItems/todoItemsSlice';

describe('RemoveButtonTest', () => {
    let store: IStoreType;

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <RemoveItemButton itemId='foo-id' color={MUIColor.Primary} />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('removes item on click', () => {
        const itemId = 'foo-id';
        const todoItemRemovedSpy = jest.spyOn(todoItemsActions, 'todoItemRemoved');

        render(
            <Provider store={store}>
                <RemoveItemButton itemId={itemId} color={MUIColor.Primary} />
            </Provider>
        );

        const toggle = screen.getByLabelText('Remove item button');
        expect(toggle).toBeInTheDocument();

        fireEvent.click(toggle);

        expect(todoItemRemovedSpy).toHaveBeenCalledTimes(1);
        expect(todoItemRemovedSpy).toHaveBeenCalledWith(itemId);
    });
});