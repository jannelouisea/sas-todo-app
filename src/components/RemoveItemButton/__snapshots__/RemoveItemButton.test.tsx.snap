// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveButtonTest renders correctly 1`] = `
<button
  aria-label="Remove item button"
  aria-labelledby={null}
  className="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorPrimary MuiIconButton-sizeSm css-sx19sf-JoyIconButton-root"
  id="remove-item-button"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ekxpud-MuiSvgIcon-root"
    data-testid="DeleteIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
    />
  </svg>
</button>
`;
